docker/scout/scripts/main_ibbtc.py

```diff
@@ -0,0 +1,281 @@
+
+ADDRESSES = checksum_address_dict(ADDRESSES_IBBTC)
+
+BLOCK_START = 12388784
+CHAIN_REORG_SAFETY_BLOCKS = 20
+POLL_INTERVAL = 60
+
+warnings.simplefilter("ignore")
+
+provider = Web3.HTTPProvider(ETHNODEURL)
+# remove the default JSON-RPC retry middleware to enable eth_getLogs block range throttling
+provider.middlewares.clear()
+w3 = Web3(provider)
+
+
+class IbbtcScannerState(BridgeScannerState):
+    def __init__(self):
+        self.state = None
+        self.fname = "ibbtc-scanner_state.json"
+        self.last_save = 0
+
+
+def process_transaction(web3, tx_hash, block_gauge, token_flow_counter, fees_counter):
+    tx = web3.eth.getTransaction(tx_hash)
+    tx_logs = web3.eth.getTransactionReceipt(tx_hash).logs
+
+    block_number = tx.blockNumber
+    block_timestamp = web3.eth.get_block(block_number)["timestamp"]
+
+    erc20_abi = json.load(open("interfaces/ERC20.json", "r"))
+    erc20 = web3.eth.contract(abi=erc20_abi)
+    erc20_transfer_abi = erc20.events.Transfer._get_event_abi()
+
+    transfers = {
+        transfer: 0
+        for transfer in [
+            "ibBTC_minted",
+            "ibBTC_burned",
+            "bcrvRenBTC_sent",
+            "bcrvRenBTC_received",
+            "bcrvSBTC_sent",
+            "bcrvSBTC_received",
+            "bcrvTBTC_sent",
+            "bcrvTBTC_received",
+            "byvWBTC_sent",
+            "byvWBTC_received",
+            "fee_badger",
+            "fee_defiDollar",
+        ]
+    }
+
+    for log in tx_logs:
+        try:
+            event_data = get_event_data(web3.codec, erc20_transfer_abi, log)
+            transfers = update_transfers(tx_hash, event_data, transfers)
+        except MismatchedABI:  # not ERC20 transfer, so skip
+            continue
+
+    # update counters
+    block_gauge.labels("block_number").set(block_number)
+    block_gauge.labels("block_timestamp").set(block_timestamp)
+
+    token_flow_counter.labels("ibBTC", "mint", "out").inc(transfers["ibBTC_minted"])
+    token_flow_counter.labels("ibBTC", "burn", "in").inc(transfers["ibBTC_burned"])
+    token_flow_counter.labels("bcrvRenBTC", "sent", "out").inc(
+        transfers["bcrvRenBTC_sent"]
+    )
+    token_flow_counter.labels("bcrvRenBTC", "received", "in").inc(
+        transfers["bcrvRenBTC_received"]
+    )
+    token_flow_counter.labels("bcrvSBTC", "sent", "out").inc(transfers["bcrvSBTC_sent"])
+    token_flow_counter.labels("bcrvSBTC", "received", "in").inc(
+        transfers["bcrvSBTC_received"]
+    )
+    token_flow_counter.labels("bcrvTBTC", "sent", "out").inc(transfers["bcrvTBTC_sent"])
+    token_flow_counter.labels("bcrvTBTC", "received", "in").inc(
+        transfers["bcrvTBTC_received"]
+    )
+    token_flow_counter.labels("byvWBTC", "sent", "out").inc(transfers["byvWBTC_sent"])
+    token_flow_counter.labels("byvWBTC", "received", "in").inc(
+        transfers["byvWBTC_received"]
+    )
+
+    fees_counter.labels("Badger DAO").inc(transfers["fee_badger"])
```

Comment:
All the flows can be handled in grafana based on functions over historical peak data (that's the power of grafan).  The only thing I see this script productivly doing is calculating fees.  There must be a better way, let's have a think about this.

In General this whole thing is meant to collect real time, point in time data, so this event based stuff isn't great and is best avoided when possible.

Reply:
Added the DefiDollar fee wallet to "badger_wallets" so its ibBTC balance should get updated along with the Badger dev multisig wallet, which is representative of the fees received.

But can't calculate total fees paid out this way if either of the wallets move any ibBTC elsewhere.